dal/elastic.go

```diff
@@ -2,126 +2,109 @@ package dal
-		log.Fatalf("quitting connection..")
+		return err
 	}
-	return client
+	e.Client = client
+	return nil
 }
 
-func ElasticGetBookByID(client *elastic.Client, id string) *elastic.GetResult{
+func (e *Elastic) GetBookByID(id string) (*models.Book, error) {
 	ctx := context.Background()
-	get1, err := client.Get().Index(INDEXNAME).Id(id).Do(ctx)
+	book, err := (e.Client).Get().Index(INDEX_NAME).Id(id).Do(ctx)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
-	return get1
+	var parsedBook models.Book
+	json.Unmarshal(book.Source, &parsedBook)
+	return &parsedBook, nil
 }
 
-func ElasticGetStoreData(client *elastic.Client) (int64, *elastic.SearchResult){
+func (e *Elastic) InsertBook(book *models.Book) (string, error) {
 	ctx := context.Background()
-	//get document count in index
-	countService := elastic.NewCountService(client)
-	countResult, err := countService.Index(INDEXNAME).Do(ctx)
+	putStatus, err := (e.Client).Index().Index(INDEX_NAME).BodyJson(*book).Do(ctx)
 	if err != nil {
-		panic(err)
-	}
-
-	//get number of distinct authors
-	agg := elastic.NewCardinalityAggregation().Field("author_name.keyword")
-	diffAuthors, err := client.Search().Index(INDEXNAME).Aggregation("diff_authors", agg).Do(ctx)
-	if err !=nil{
-		panic(err)
+		return "", err
 	}
-	return countResult,diffAuthors
-
+	return putStatus.Id, nil
 }
-
-func ElasticPutBook(client *elastic.Client, book *Book) *elastic.IndexResponse{
+func (e *Elastic) UpdateBook(title string, id string) error {
 	ctx := context.Background()
-	// Index a tweet (using JSON serialization)
-	put1, err := client.Index().Index(INDEXNAME).BodyJson(*book).Do(ctx)
+	_, err := (e.Client).Update().Index(INDEX_NAME).Id(id).Doc(map[string]interface{}{"title": title}).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return err
 	}
-	return put1
+	return nil
 }
-
-func ElasticUpdateBook(client *elastic.Client, title string, id string) *elastic.UpdateResponse{
+func (e *Elastic) DeleteBook(id string) error {
 	ctx := context.Background()
-	response, err := client.Update().Index(INDEXNAME).Id(id).Doc(map[string]interface{}{"title": title}).Do(ctx)
+	_, err := (e.Client).Delete().Index("books_yuval").Id(id).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return err
 	}
-	return response
+	return nil
 }
 
-func ElasticDeleteBook(client *elastic.Client, id string) *elastic.DeleteResponse{
+func (e *Elastic) SearchBook(title string, authorName string, priceRangeStr string) ([]models.Book, error) {
 	ctx := context.Background()
-	response, err := client.Delete().Index("books_yuval").Id(id).Do(ctx)
+	query := elastic.NewBoolQuery()
+	if FieldExist(title) {
+		matchQuery1 := elastic.NewMatchQuery("title", title)
+		query.Filter(matchQuery1)
+	}
+	if FieldExist(authorName) {
+		matchQuery2 := elastic.NewMatchQuery("author_name", authorName)
+		query.Filter(matchQuery2)
+	}
+	if FieldExist(priceRangeStr) {
+		priceRange := GetPriceRange(priceRangeStr)
```

Comment:
The only place that uses this function is here - so there is no need to put this function in the `utils` file.

and - if there is an error, you still want to serve the request? or do you want to tell the user that he used an incorrect input?